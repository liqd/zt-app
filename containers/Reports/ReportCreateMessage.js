import React from 'react';
import { Alert, View, ScrollView } from 'react-native';
import { Button } from '@rneui/base';
import { Formik } from 'formik';
import * as yup from 'yup';
import IconSLI from 'react-native-vector-icons/SimpleLineIcons';
import { ButtonSubmit } from '../../components/ButtonSubmit';
import { TextSourceSans } from '../../components/TextSourceSans';
import AsyncStorage from '@react-native-async-storage/async-storage';
import API from '../../BaseApi';

import { styles } from './ReportCreateMessage.styles';
import { TextInputFullFormField } from '../../components/formFields';
import { VirtualScrollView } from '../../components/VirtualScrollView';

export const ReportCreateMessage = props => {
  const {content_type, object_pk} = props.route.params;

  const arrowLeftIcon = (
    <IconSLI name='arrow-left' size={22} />
  );
  const reportMessageValidationSchema = yup.object().shape({
    message: yup
      .string()
      .max(1024, 'Message must be no longer then 1024 characters')
  });

  const handleSubmit = (values) => {
    const description = values.message;
    AsyncStorage.getItem('authToken')
      .then((token) => {
        return API.postReport({content_type, object_pk, description}, token);
      })
      .then((response) => {
        const {statusCode, data} = response;
        if (statusCode == 201) {
          Alert.alert('Thank you! We are taking care of it.');
          props.navigation.goBack();
        }
        else {
          const errorMessage = 'That did not work.';
          let errorDetail;
          if (statusCode==403) {
            errorDetail = data.detail;
          }
          else if (statusCode == 400) {
            errorDetail = 'Bad request';
          }
          Alert.alert(errorMessage, errorDetail);
        }
      });
  };

  return (
    <View style={styles.container}>
      <View style={styles.actionsContainer}>
        <Button
          buttonStyle={styles.backButton}
          titleStyle={styles.backButtonText}
          title='Back'
          type='clear'
          icon={arrowLeftIcon}
          onPress={() => props.navigation.goBack()}
        />
      </View>
      <Formik
        validationSchema={reportMessageValidationSchema}
        initialValues={{message: ''}}
        onSubmit={values => handleSubmit(values)}
        validateOnMount={true}
      >
        {({
          handleChange,
          handleBlur,
          handleSubmit,
          values,
          errors,
          touched,
          isValid,
        }) => (
          <>
            <ScrollView>
              <TextSourceSans style={styles.title}>
                You want to report this content? Your message will be sent to the
                moderation. The moderation will look at the reported content. The
                content will be deleted if it does not meet our discussion rules
                (netiquette).
              </TextSourceSans>
              <TextInputFullFormField
                name='message'
                value={values.message}
                placeholder='Add message'
                returnKeyType='next'
                returnKeyLabel='next'
                onChangeText={handleChange('message')}
                onBlur={handleBlur('message')}
                error={errors.message}
                touched={touched.message}
                numberOfLines={10}
                multiline={true}
              />
            </ScrollView>
            <ButtonSubmit
              title='Submit'
              onPress={handleSubmit}
              disabled={!isValid}
            >
            </ButtonSubmit>
          </>
        )}
      </Formik>
    </View>
  );
};

ReportCreateMessage.navigationOptions = {
  headerTitle: 'Add your message',
  // headerBackTitle only for iOS
  headerBackTitle: 'Back'
};
