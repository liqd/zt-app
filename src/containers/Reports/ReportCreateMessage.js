import React, { useEffect, useState } from 'react'
import { Alert, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import { Formik } from 'formik'
import * as yup from 'yup'

import API from '../../BaseApi'
import { TextInputFullFormField } from '../../components/formFields'
import { Header } from '../../components/Header'
import { KeyboardScrollView } from '../../components/KeyboardScrollView'
import { TextSourceSans } from '../../components/TextSourceSans'

import { styles } from './ReportCreateMessage.styles'

export const ReportCreateMessage = props => {
  const {content_type, object_pk} = props.route.params
  const [error, setError] = useState()
  const [submitPending, setSubmitPending] = useState()

  const reportMessageValidationSchema = yup.object().shape({
    message: yup
      .string()
      .max(1024, 'Message must be no longer then 1024 characters')
      .required('Please add a few words explaining why you are reporting this content'),
  })

  useEffect(() => {
    if (error) {
      Alert.alert('An error occured', error, [{ text: 'Ok' }])
    }
  }, [error])

  const handleSubmit = (values) => {
    const description = values.message
    setSubmitPending(true)
    return API.postReport({
      content_type,
      object_pk,
      description
    })
      .then((response) => {
        const {statusCode, data} = response
        if (statusCode === 201) {
          Alert.alert('Thank you! We are taking care of it.')
          setSubmitPending(false)
          props.navigation.goBack()
        } else if (statusCode === 403) {
          return Promise.reject(data.detail)
        } else if (statusCode === 400) {
          return Promise.reject('Bad request')
        }
      })
      .catch(error => {
        setError(error)
        setSubmitPending(false)
      })
  }

  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.header}>
        <Header navigation={props.navigation} style={styles.header}/>
      </View>
      <Formik
        validationSchema={reportMessageValidationSchema}
        initialValues={{message: ''}}
        onSubmit={values => handleSubmit(values)}
        validateOnMount={true}
      >
        {({
          handleChange,
          handleBlur,
          handleSubmit,
          values,
          errors,
          touched,
          isValid,
        }) => (
          <KeyboardScrollView
            handleSubmit={handleSubmit}
            isValid={isValid && !submitPending}
            buttonText='Submit'
          >
            <TextSourceSans style={styles.title}>
              You want to report this content? Your message will be sent to the
              moderation. The moderation will look at the reported content. The
              content will be deleted if it does not meet our discussion rules
              (netiquette).
            </TextSourceSans>
            <TextInputFullFormField
              name='message'
              value={values.message}
              placeholder='Add message'
              returnKeyType='next'
              returnKeyLabel='next'
              onChangeText={handleChange('message')}
              onBlur={handleBlur('message')}
              error={errors.message}
              touched={touched.message}
              numberOfLines={10}
              multiline={true}
            />
          </KeyboardScrollView>
        )}
      </Formik>
    </SafeAreaView>
  )
}

ReportCreateMessage.navigationOptions = {
  headerTitle: 'Add your message',
  // headerBackTitle only for iOS
  headerBackTitle: 'Back'
}
