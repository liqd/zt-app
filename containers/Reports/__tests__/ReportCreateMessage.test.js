import React from 'react'
import { fireEvent,render, screen, waitFor } from '@testing-library/react-native'

import '@testing-library/jest-native/extend-expect'

import { testIdeaAdmin } from '../../../tests/TestData'
import { ReportCreateMessage } from '../ReportCreateMessage'

test('Test ReportCreateMessage Snapshot', () => {
  const route = { params: {
    content_type: testIdeaAdmin.content_type,
    object_pk: testIdeaAdmin.pk
  }}
  const { toJSON } = render(<ReportCreateMessage />)
  expect(toJSON()).toMatchSnapshot()
})

test('Send report, too many chars', async () => {
  render(<ReportCreateMessage />)
  const submitButton = screen.getByRole('button')
  fireEvent(submitButton, 'press')
  const validationErrors = await screen.findAllByText(/Message must be no longer then 1024 characters/)
  expect(validationErrors).toBeTruthy()
})
