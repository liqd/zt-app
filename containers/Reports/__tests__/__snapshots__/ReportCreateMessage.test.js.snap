// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ReportCreateMessage Snapshot 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "paddingTop": 5,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#fff",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingBottom": 20,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 2,
              },
              undefined,
              false,
            ]
          }
          testID="RNE_BUTTON_WRAPPER"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "busy": false,
                "disabled": false,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderColor": "#2089dc",
                  "borderRadius": 2,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "marginVertical": 0,
                  "padding": 8,
                  "paddingHorizontal": 10,
                }
              }
            >
              <Text />
              <Text
                style={
                  Object {
                    "color": "#1a1a1a",
                    "fontSize": 17,
                    "marginHorizontal": 5,
                    "paddingVertical": 1,
                    "textAlign": "center",
                  }
                }
              >
                Back
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
          "justifyContent": "space-between",
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "SourceSansPro_400Regular",
              },
              Object {
                "fontSize": 17,
                "lineHeight": 20,
                "paddingHorizontal": 15,
              },
            ]
          }
        >
          You want to report this content? Your message will be sent to the moderation. The moderation will look at the reported content. The content will be deleted if it does not meet our discussion rules (netiquette).
        </Text>
        <View
          style={
            Object {
              "paddingHorizontal": 15,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "SourceSansPro_400Regular",
                },
                Object {
                  "fontSize": 15,
                  "lineHeight": 19,
                  "paddingBottom": 5,
                  "paddingTop": 10,
                  "width": "100%",
                },
              ]
            }
          />
          <TextInput
            accessible={true}
            multiline={true}
            name="message"
            numberOfLines={10}
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder="Add message"
            returnKeyLabel="next"
            returnKeyType="next"
            style={
              Object {
                "alignContent": "flex-start",
                "fontSize": 17,
                "justifyContent": "flex-start",
                "textAlignVertical": "top",
              }
            }
            value=""
          />
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 2,
          },
          undefined,
          false,
        ]
      }
      testID="RNE_BUTTON_WRAPPER"
    >
      <View
        accessibilityLabel="Submit"
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": false,
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#2d40cc",
              "borderColor": "#2089dc",
              "borderRadius": 0,
              "borderWidth": 0,
              "flexDirection": "row",
              "fontSize": 16,
              "height": 70,
              "justifyContent": "center",
              "padding": 8,
              "paddingHorizontal": 10,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            Submit
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
